### PR TITLE
Enable L0 hyperparameter tuning

### LINKED ISSUE #76: Create target holdout validation function

### LINKED ISSUE #77: Implement optuna l0 hyperparameter tuning

### PR DESCRIPTION
**Pull request by juaristi22**
Fix #77 
Fix #76 

This enables tuning the L0 parameters to achieve a combination of loss reduction, estimates being within 10% of targets, and sparsity (proportion of weights that are 0). The objective function combines the three, and the parameter "objectives_contribution" allows changing how much we want to prioritize each objective, making it flexible to tune for different needs. 

It also supports target holdout. `n_holdout_sets` can be set to 1 with a `holdout_fraction` of 0 if we dont want to holdout targets.

### CODE REVIEW

**Review by baogorek — CHANGES_REQUESTED** [review]
Hi Maria, I just ran the test and it ran very smoothly and I saw the sparse output. I see that you were able to leverage an open-source parameter optimizer (Optuna), which was smart. One thing I would like to see is a bit of information about the holdouts once tuning is one. There should be an error for each of the holdout sets and a way to see which targets were held out. I'm probably not going to review every line of code here, but if I could see that, I would feel very good about putting my checkmark here.

File: src/microcalibrate/calibration.py [hunk 1]
```diff
@@ -463,3 +487,916 @@ def summary(
...
+        Returns:
+            List of dictionaries containing holdout names and indices
+        """
+        n_targets = len(self.target_names)
+        n_holdout_targets = max(1, int(n_targets * holdout_fraction))
+
+        holdout_sets = []
+        for i in range(n_holdout_sets):
+            # Each holdout set gets a different random selection
+            set_rng = np.random.default_rng((random_state or self.seed) + i)
```
Comment (baogorek) [inline src/microcalibrate/calibration.py:920]: I'm looking at the holdout sets created by the test:

```
In [30]: holdout_sets
Out[30]: 
[{'names': ['total_population', 'young_employed'], 'indices': array([0, 6])},
 {'names': ['income_west', 'income_south'], 'indices': array([5, 3])},
 {'names': ['income_east', 'total_population'], 'indices': array([4, 0])}]
```
 
In the first set, there's young_employed` in the holdout, but then that implies that employed_count` will be in the training set, since:

```
In [41]: self.target_names
Out[41]: 
array(['total_population', 'employed_count', 'income_north',
       'income_south', 'income_east', 'income_west', 'young_employed',
       'senior_count'], dtype=object)
```
It's at least partial leakage of information from the training fold to the holdout fold. We don't have a target concept right now, so we may just want to throw a warning that data leakage is almost guaranteed to happen at some level.

File: src/microcalibrate/calibration.py [hunk 2]
```diff
@@ -463,3 +487,916 @@ def summary(
...
+                self.target_names.copy()
+                if self.target_names is not None
+                else None
+            ),
+        }
+
+        # Initialize list to collect all holdout evaluations
+        all_evaluations = []
+
+        def evaluate_single_holdout(
```
Comment (baogorek) [inline src/microcalibrate/calibration.py:570]: Pretty complex function to be defined within another function. Feels like this could be standalone logic.

File: src/microcalibrate/calibration.py [hunk 3]
```diff
@@ -463,3 +487,916 @@ def summary(
         ) / df["Official target"]
         df = df.reset_index(drop=True)
         return df
+
+    def tune_hyperparameters(
```
Comment (baogorek) [inline src/microcalibrate/calibration.py:491]: Given how large this function is, and that there are functions within function, I wonder if it would make sense for it to be its own module.

File: src/microcalibrate/calibration.py [hunk 4]
```diff
@@ -463,3 +487,916 @@ def summary(
...
+                    i for i in range(n_targets) if i not in val_indices
+                ]
+
+                val_estimates = all_estimates[val_indices]
+                val_targets = self.original_targets[val_indices]
+                train_estimates = all_estimates[train_indices]
+                train_targets = self.original_targets[train_indices]
+
+                # Calculate metrics
+                from .utils.metrics import loss, pct_close
```
Comment (baogorek) [inline src/microcalibrate/calibration.py:646]: If you wouldn't mind, spelling out `from microcalibrate.utils.metrics import loss, pct_close` makes it easier for me to run interactively. Also, the AIs are usually telling me to declare my imports at the top. Is it really doing much for us to delay this import until now (I can see it for `from policyengine_us import Microsimulation`, because it takes like 7 seconds).

**Review by baogorek — APPROVED** [review]
There's a lot in here, and I don't want to block useful functionality from going in. I am worried about data leakage from the training folds into the holdout folds, which is virtually guaranteed to happen given the way we define targets. All I would ask is that there be some warning added so the user can see it and we will be reminded to make it more robust in the future.
